libyottadb/SimpleAPI.rst

```diff
@@ -0,0 +1,1010 @@
+(Atomic, Consistent, Isolated, Durable) transactions
+<https://en.wikipedia.org/wiki/ACID>`_, ensuring strict serialization
+of transactions, using `optimistic concurrency control
+<http://sites.fas.harvard.edu/~cs265/papers/kung-1981.pdf>`_.
+
+Here is a simplified view [#]_ of YottaDB's implementation of
+optimistic concurrency control:
+
+- Each database file header has a field of the next *transaction
+  number* for updates in that database.
+- The block header of each database block in a database file has the
+  transaction number when that block was last updated.
+- When a process is inside a transaction, it keeps track of every
+  database block it has read, and the transaction numbner of that
+  block when read. Other processes are free to update the database
+  during this time.
+- The process retains updates in its memory, without committing them
+  to the database, so that it's own logic sees the updates, but no
+  other process does. As every block that the process wishes to write
+  must also be read, tracking the transaction numbers of blocks read
+  suffices to track them for blocks to be writen.
+- To commit a transaction, a process checks whether any block it has
+  read has been updated since it was read. If none has, the process
+  commits the transaction to the database, incrementing the file
+  header fields of each updated database file for the next
+  transaction.
+- If even one block has been updated, the process discards its work,
+  and starts over. If after three attempts, it is still unable to
+  commit the transaction, it executes the transaction logic on the
+  fourth attempt with updates by all other processes blocked so that
+  the transaction at commit time will not encounter database changes
+  made by other processes.
+
+.. [#] At the high level at which optimistic concurrency control is
+       described here, a single logical database update (which can
+       span multiple blocks and even multiple regions) is a
+       transaction that contains a single update.
+
+In libyottadb's API for transaction processing, an application
+packages the logic for a transaction into a function with one
+parameter, passing the function and its parameter as parameters to the
+`ydb_tp_s()`_ function. libyottadb then calls that function.
+
+- If the function returns a ``YDB_OK``, libyottadb attempts to commit
+  the transaction. If it is unable to commit as described above, or if
+  the called function returns a ``YDB_TP_RESTART`` return code, it
+  calls the function again.
+- If the function returns a ``YDB_TP_ROLLBACK``, `ydb_tp_s()`_ returns
+  to its caller with that return code.
+- To protect applications against poorly coded transactions, if a
+  transaction takes longer than the number of seconds specified by
+  the environment variable ``ydb_maxtptime``, libyottadb aborts the
+  transaction and the `ydb_tp_s()`_ function returns the
+  ``YDB_ERR_TPTIMEOUT`` error.
+
```

Comment:
If the transaction is stuck in C code (as opposed to M code), then we might not be able to abort it until the next M statement is reached. Does that need mention here?